### PR TITLE
Make shadowJar task cacheable

### PR DESCRIPTION
**Pull request by ghale**
These changes make the `shadowJar` task compatible with the build cache.  

The task will still be cacheable if any of the following transformers are used:

- AppendingTransformer
- XmlAppendingTransformer
- ServiceFileTransformer
- GrroovyExtensionModuleTransformer

But the other out-of-the-box transformers are not cacheable yet (if they are added, the task will be marked as un-cacheable) but I can add caching for those with a future PR.  User-defined transformers will also make the task un-cacheable unless they declare the inputs of the transformer and annotate it with `@CacheableTransformer`.

Similarly, `SimpleRelocator` has been marked as cacheable, but any custom relocators will disable caching unless the class is annotated with `@CacheableRelocator`.

We've seen fairly significant task execution times for projects with large dependency graphs, so making this task usable with the build cache could provide a decent performance improvement.

Please pay special attention to the test coverage and let me know if there are any potential edge cases not covered.  I'm happy to discuss anything if it doesn't make sense or seems wrong.

### CODE REVIEW

File: src/main/groovy/com/github/jengelman/gradle/plugins/shadow/relocation/SimpleRelocator.groovy [hunk 1]
```diff
@@ -188,4 +190,39 @@ class SimpleRelocator implements Relocator {
...
+
+    @Input
+    String getShadedPattern() {
+        return shadedPattern
+    }
+
+    @Input
+    String getShadedPathPattern() {
+        return shadedPathPattern
+    }
```
Comment (Goooler) [inline src/main/groovy/com/github/jengelman/gradle/plugins/shadow/relocation/SimpleRelocator.groovy:212]: @ghale

These properties are immutable and unused, can we remove them? I'm planning to remove them and migrate `includes` and `excludes` to `org.gradle.api.provider.SetProperty`.

Comment (ghale) [reply src/main/groovy/com/github/jengelman/gradle/plugins/shadow/relocation/SimpleRelocator.groovy:212]: Yeah, if these are being removed in favor of property objects, then yes, these can be removed.  They were only added in order to expose the inputs from the relocator.

Comment (ghale) [reply src/main/groovy/com/github/jengelman/gradle/plugins/shadow/relocation/SimpleRelocator.groovy:212]: To be clear, the new property objects should also be exposed as inputs (i.e. a getter with an `@Input` annotation) in order to maintain proper cacheability after these are removed.

Comment (Goooler) [reply src/main/groovy/com/github/jengelman/gradle/plugins/shadow/relocation/SimpleRelocator.groovy:212]: Can you take a review for #1079?

File: src/main/groovy/com/github/jengelman/gradle/plugins/shadow/transformers/CacheableTransformer.groovy [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+import java.lang.annotation.Target
+
+/**
+ * Marks that a given instance of {@link Transformer} is is compatible with the Gradle build cache.
+ * In other words, it has its appropriate inputs annotated so that Gradle can consider them when
+ * determining the cache key.
+ */
+@Retention(RetentionPolicy.RUNTIME)
+@Target(ElementType.TYPE)
+@interface CacheableTransformer {
```
Comment (Goooler) [inline src/main/groovy/com/github/jengelman/gradle/plugins/shadow/transformers/CacheableTransformer.groovy:15]: @ghale 

I mark all built-in transformers cacheable, it would be great if you can take a review for #1210.
